daml-assistant/src/DA/Daml/Assistant/Command.hs

```diff
@@ -55,26 +57,44 @@ builtin name desc mod parser =
 isHidden :: SdkCommandInfo -> Bool
 isHidden = isNothing . sdkCommandDesc
 
-dispatch :: SdkCommandInfo -> Mod CommandFields Command
-dispatch info = subcommand
+dispatch :: Env -> SdkCommandInfo -> Mod CommandFields Command
+dispatch env info = subcommand
     (unwrapSdkCommandName $ sdkCommandName info)
     (fromMaybe "" $ sdkCommandDesc info)
     forwardOptions
-    (Dispatch info . UserCommandArgs <$>
-        many (strArgument (metavar "ARGS" <> completer defaultCompleter)))
-
-commandParser :: [SdkCommandInfo] -> Parser Command
-commandParser cmds | (hidden, visible) <- partition isHidden cmds = asum
+    (Dispatch info . UserCommandArgs <$> sdkCommandArgsParser env info)
+
+sdkCommandArgsParser :: Env -> SdkCommandInfo -> Parser [String]
+sdkCommandArgsParser env info = fromM (go (unwrapSdkCommandArgs $ sdkCommandArgs info))
+  where go args = do
+          mx <- oneM (optional $ strArgument (completer $ nestedCompl args))
+          case mx of
+            Nothing -> return []
+            Just x -> (x :) <$> go (args ++ [x])
+        nestedCompl args = mkCompleter $ \arg -> do
+          let path = unwrapSdkPath (fromJust $ envSdkPath env) </> unwrapSdkCommandPath (sdkCommandPath info)
```

Comment:
It looks like the only reason you need `Env` here is to calculate the absolute path of the SDK command. I think I'd prefer this to be part of the `SdkCommandInfo` structure, instead of passing the whole `Env` around here. (And it gets rid of the `error` call without making things more complicated here.)

You need to update the `SdkCommandInfo` in any case, to handle turning completions on or off via sdk-config.yaml.

Reply:
Makes sense :+1: